doc/amtu.8
.\" Copyright (C) 2003 IBM
.\" This file is distributed according to the GNU General Public License.
.\" See the file COPYING in the top level source directory for details.
.de Sh \" Subsection
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp 
.if t .sp .5v
.if n .sp
..
.de Ip \" List item
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.TH "AMTU" 8 "2003-08-25" "Linux 2.4" "Linux System Administration"
.SH NAME
amtu \- Abstract Machine Test Utility (AMTU)
.SH "SYNOPSIS"

.nf
\fBamtu\fR [\fB-dmsinph\fR]
.fi

.SH "DESCRIPTION"

.PP
Abstract Machine Test Utility (AMTU) is an administrative utility that 
checks whether the underlying protection mechanisms of the hardware are 
being enforced. 
These checks are a requirement of the Controlled Access Protection Profile 
(CAPP) FPT_AMT.1 
(see http://www.radium.ncsc.mil/tpep/library/protection_profiles/CAPP-1.d.pdf).

.PP
AMTU executes the following tests:

.TP
* Memory
Randomly writes to areas of memory, then reads the memory back to ensure 
the values written remain unchanged.

.TP
* Memory Separation
Ensures that user space programs cannot read and write to areas of memory 
utilized by items such as Video RAM and kernel code.


.TP
* I/O Controller - Network
Verifies random data transmitted is also the data received for each 
configured network device. 
Checks only Ethernet and token ring devices that are configured and up. 
Does not check async devices.

.TP
* I/O Controller - Disk
Verifies random data written to disks remains unchanged. 
Only IDE and SCI controllers that are associated with mounted file systems 
are checked. 
Disk controllers with read-only mounted file systems are not checked.

.TP
* Supervisor Mode Instructions
Ensures that the enforcement of the property that privileged instructions 
should only be in supervisor mode is still in effect. 
The set of privileged instructions tested to confirm this is 
architecture dependent.

.SH "OPTIONS"

.PP
Options available for the \fBamtu\fR command are the following:

.TP
\fB-d\fR
Print debug messages.

.TP
\fB-m\fR
Execute Memory Test.

.TP
\fB-s\fR
Execute Memory Separation Test.

.TP
\fB-i\fR
Execute I/O Controller - Disk Test.

.TP
\fB-n\fR
Execute I/O Controller - Network Test

.TP
\fB-p\fR
Execute Supervisor Mode Instructions Test.

.TP
\fB-h\fR
Print help message.

.SH "RETURN CODES"

.PP
The \fBamtu\fR command issues the following return codes when executed:

.TP
-1
Failure

.TP
0
Success.

.SH "REPORTING BUGS"

.PP
Report bugs to amtueal-devel@lists.sourceforge.net .

.SH AUTHOR
Emily Ratliff, Joy Latten, and Kimberly Simon.


src/amtu.c
//----------------------------------------------------------------------
//
// Module Name:  amtu.c
//
// Include File:  amtu.h
//
// Description:   Code for Abstract Machine Test Utility.
//
// Notes:  This module is the framework for the Abstract Machine
//         Test Utility (AMTU). The following tests will be invoked
//         from this module:
//         - Memory
//         - Memory Separation
//         - I/O Controller - Network
//         - I/O Controller - Disk
//         - Supervisor Mode Instructions
// -----------------------------------------------------------------
// LANGUAGE:     C
//
// (C) Copyright IBM Corp. 2003,2005
// Licensed under the Common Public License
// -----------------------------------------------------------------
//
// Change Activity:
// DATE     PGMR      COMMENTS
// ------- --------- ----------------------
// 07/29/03 K.Simon  Added iodisktest function 
// 08/25/03 K.Simon  Added audit tag for LAUS_LOG  		      
// 08/26/03 K.Simon  Improved usage statement
// 10/17/03 K.Simon  Removed NO_TAG
// 10/20/03 K.Simon  Added msg variable to allow LAUS_LOG
//		     to print variables
// 03/15/05 D.Velarde Added AUDIT_LOG and AUDIT_OPEN statements to be
//                    used if LIBLAUS isn't being used
// 03/16/05 D.Velarde Stop auditing when test is done.
// 07/12/05 D.Velarde Do NOT start or stop audit daemon.  Remove all
//                    AUDIT_OPEN and AUDIT_CLOSE calls.
// 12/08/05 D.Velarde Updated with new records patch from RedHat
//
//----------------------------------------------------------------------

#include "config.h"
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <syscall.h>
#include <errno.h>
#include <string.h>
#include <signal.h>
#include <syslog.h>
#include <sys/types.h>
#include <sys/wait.h>
#include "amtu.h"

void usage()
{
	printf("Usage: amtu [-dmsinph]\n");
	printf("d      Display debug messages\n");
	printf("m      Execute Memory Test\n");
	printf("s      Execute Memory Separation Test\n");
	printf("i      Execute I/O Controller - Disk Test\n");
	printf("n      Execute I/O Controller - Network Test\n");
	printf("p      Execute Supervisor Mode Instructions Test\n");
	printf("h      Display help message\n");
	exit(-1);
}

int main(int argc, char *argv[])
{
	int rc = 0;
	int c;
	int memtest = 0, memseptest = 0, disktest = 0;
	int nettest = 0, privtest = 0;
	int testspecified = 1;
	char msg[50];

#ifdef HAVE_LIBLAUS
	LAUS_OPEN
#endif
	
	while ((c = getopt(argc, argv, "dmsinph")) != -1) { 
		switch (c) {
			case 'd':
				debug = 1;
				break;
			case 'm':
				memtest++;
				testspecified = 0;
				break;
			case 's':
				memseptest++;
				testspecified = 0;
				break;
			case 'i':
				disktest++;
				testspecified = 0;
				break;
			case 'n':
				nettest++;
				testspecified = 0;
				break;
			case 'p':
				privtest++;
				testspecified = 0;
				break;
			case 'h':
				usage();
				break;
			default:
				usage();
				break;
		}
	}

	// Invoke Memory Test
	if (testspecified || memtest) {
		rc = memory(argc, argv); 
	}

	// Invoke Memory Separation Test
	if (testspecified || memseptest) {
		rc |= memsep(argc, argv);
	}

	// Invoke I/O Controller - Network Test
	if (testspecified || nettest) {
		rc |= networkio(argc, argv);
	}	

	// Invoke I/O Controller - Disk Test
	if (testspecified || disktest) {
		rc |= iodisktest(argc, argv);
	}

	// Invoke Supervisor Mode Instructions Test
	if (testspecified || privtest) {
		rc |= amtu_priv(argc, argv);
	}

	if (rc != 0) {
		sprintf(msg, "amtu failed: %d", rc);
#ifdef HAVE_LIBLAUS
		LAUS_LOG((msg))
#else
		AUDIT_LOG(msg, 0)
#endif
	} else {
#ifdef HAVE_LIBLAUS
		LAUS_LOG(("amtu completed successfully."))
#else
		AUDIT_LOG("amtu completed successfully", 1)
#endif
	}
	return(rc);
}
